### PR TITLE
Add Typescript Postgres UserService implementation

### PR DESCRIPTION
**Pull request by zm2he**
## Notion ticket link

[Firebase authentication & user services using PostgreSQL (Sequelize)](https://www.notion.so/uwblueprintexecs/Firebase-authentication-user-services-using-PostgreSQL-Sequelize-3338837f5b4c44b4a011ba60339f6403)

## Implementation description
* Implemented methods defined in `IUserService` in temporary file `userServicepg.ts` with  Firebase Admin SDK for auth related user management logic
* Defined Postgres `User` using Sequelize in `user.pgmodel.ts`

## Steps to test
Testing code in `server.ts` will be added shortly.

## What should reviewers focus on?
* Most important file is `backend/typescript/services/implementations/userServicepg.ts`
* Delete and Update are more substantially different from the MongoDB implementation

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/services/implementations/userServicepg.ts [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+
+class PGUserService implements IUserService {
+  /* eslint-disable class-methods-use-this */
+
+  async getUserById(userId: string): Promise<UserDTO> {
+    let user: User | null;
+    let firebaseUser: firebaseAdmin.auth.UserRecord;
+
+    try {
+      user = await User.findByPk(Number(userId));
```
Comment (sherryhli) [inline backend/typescript/services/implementations/userServicepg.ts:15]: No change required but converting to number is not needed

File: backend/typescript/services/implementations/userServicepg.ts [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+      throw error;
+    }
+  }
+
+  async deleteUserByEmail(email: string): Promise<void> {
+    try {
+      const firebaseUser: firebaseAdmin.auth.UserRecord = await firebaseAdmin
+        .auth()
+        .getUserByEmail(email);
+      const deletedUser: User | null = await User.findByPk(Number(firebaseUser.userId));
```
Comment (sherryhli) [inline backend/typescript/services/implementations/userServicepg.ts:296]: Should be finding by `authId == firebaseUser.uid`?

```ts
const deletedUser: User | null = await User.findOne({ 
  where: { authId: firebaseUser.uid },
});
```

**Review by sherryhli — CHANGES_REQUESTED** [review]
A couple changes, also please run `yarn fix` again

File: backend/typescript/services/implementations/userServicepg.ts [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+      }
+
+      User.update({
+        firstName: user.firstName,
+        lastName: user.lastName,
+        role: user.role
+       },
+       {
+        where: { id: userId}
+       });
```
[the remark this answers is not in the document]
Comment (zm2he) [reply backend/typescript/services/implementations/userServicepg.ts:204]: I believe User.update gives you the number of rows affected and the number of each affected row, not the actual user if I'm not mistaken.

File: backend/typescript/models/user.pgmodel.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { Column, DataType, Model, Table } from "sequelize-typescript";
+import { Role } from "../types";
+
+@Table({ tableName: "User" })
```
[the remark this answers is not in the document]
Comment (zm2he) [reply backend/typescript/models/user.pgmodel.ts]: Fixed.

File: backend/typescript/services/implementations/userServicepg.ts [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+          lastName: user.lastName,
+          authId: firebaseUser.uid,
+          role: user.role,
+        });
+      } catch (postgresError) {
+        try {
+          await firebaseAdmin.auth().deleteUser(firebaseUser.uid);
+        } catch (firebaseError) {
+          const errorMessage = [
+            "Failed to rollback Firebase user creation after MongoDB user creation failure. Reason =",
```
[the remark this answers is not in the document]
Comment (zm2he) [reply backend/typescript/services/implementations/userServicepg.ts]: Fixed.

File: backend/typescript/services/implementations/userServicepg.ts [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+  }
+
+  async getUserByEmail(email: string): Promise<UserDTO> {
+    let user: User | null;
+    let firebaseUser: firebaseAdmin.auth.UserRecord;
+
+    try {
+      firebaseUser = await firebaseAdmin.auth().getUserByEmail(email);
+      user = await User.findOne({ 
+        where: { authId: firebaseUser.authId },
```
[the remark this answers is not in the document]
Comment (zm2he) [reply backend/typescript/services/implementations/userServicepg.ts]: Fixed.

File: backend/typescript/services/implementations/userServicepg.ts (hunk 2 above)
Comment (zm2he) [reply backend/typescript/services/implementations/userServicepg.ts]: Fixed.

File: backend/typescript/services/implementations/userServicepg.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (sherryhli) [reply backend/typescript/services/implementations/userServicepg.ts]: The return type is a 2-tuple of [affected row count (int), affected values (array of users)], so this should work. We've also tested it out on the entity service :)

**Review by sherryhli — COMMENTED** [review]
1 final comment to address here, otherwise good to go
https://github.com/uwblueprint/starter-code-v2/pull/16#discussion_r618770092

**Review by sherryhli — APPROVED** [review]

**Pull request merged**